### PR TITLE
Chore/introduce closed factories

### PR DESCRIPTION
**Pull request by DominikGuzei**
Adds closed factory (see [Factories Wiki Page](https://github.com/stravid/slotcars/wiki/Factories) and changes screen factory to be part of the game! Lets all screens register themselves at the factory and thus removes a lot of dependencies from the screen factory. Adding a new screen will just require changes in the new screen and the slotcars application code.

The `ScreenFactory` extends `Factory` and can be accessed in the code with the static `ScreenFactory.get`method which returns the singleton instance.

### CODE REVIEW

File: app/assets/javascripts/slotcars/factories/factory.js.coffee [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  registerType: (id, type) -> @_registeredTypes[id] = type
+
+  getInstanceOf: (typeId, createParamters={}) ->
+    if @_registeredTypes[typeId]?
+      @_registeredTypes[typeId].create createParamters
+    else
+      throw "#{typeId} was not registered in factory."
+
+Factory.reopenClass
```
Comment (tombu) [inline app/assets/javascripts/slotcars/factories/factory.js.coffee:17]: please comment why you use `reopenClass`

File: app/assets/javascripts/slotcars/factories/factory.js.coffee [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  instance: null
+
+  get: -> if @instance? then @instance else @instance = @create()
```
Comment (tombu) [inline app/assets/javascripts/slotcars/factories/factory.js.coffee]: I prefer `getInstance()` instead of just `get()` ... it´s a matter of taste

**Comment by DominikGuzei** [comment]
I just realized that the slotcars application can be simplified to just one method `showScreen` that takes a string and does the same stuff thats happening now in each separate method for each screen. This would remove even more unnecessary code duplication :-D

**Comment by tombu** [comment]
do it, then ship it :shipit:

**Comment by stravid** [comment]
Looks good!

**Pull request closed without merging**